crates/sui-kvstore/src/lib.rs

```diff
@@ -0,0 +1,52 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+mod bigtable;
+use anyhow::Result;
+use async_trait::async_trait;
+pub use bigtable::client::BigTableClient;
+use sui_types::base_types::SequenceNumber;
+use sui_types::digests::TransactionDigest;
+use sui_types::effects::{TransactionEffects, TransactionEvents};
+use sui_types::full_checkpoint_content::CheckpointData;
+use sui_types::messages_checkpoint::{
+    CheckpointContents, CheckpointSequenceNumber, CheckpointSignatureMessage, CheckpointSummary,
+};
+use sui_types::object::Object;
+use sui_types::storage::ObjectKey;
+use sui_types::transaction::Transaction;
+
+#[async_trait]
+pub trait KeyValueStoreReader {
+    async fn get_objects(&mut self, objects: &[ObjectKey]) -> Result<Vec<Object>>;
+    async fn get_transactions(
+        &mut self,
+        transactions: &[TransactionDigest],
+    ) -> Result<Vec<TransactionData>>;
+    async fn get_checkpoint(
+        &mut self,
+        sequence_number: SequenceNumber,
+    ) -> Result<Option<Checkpoint>>;
+}
+
+#[async_trait]
+pub trait KeyValueStoreWriter {
+    async fn save_objects(&mut self, objects: &[&Object]) -> Result<()>;
+    async fn save_transactions(&mut self, transactions: &[TransactionData]) -> Result<()>;
+    async fn save_checkpoint(&mut self, checkpoint: &CheckpointData) -> Result<()>;
+}
+
+#[derive(Clone, Debug)]
+pub struct Checkpoint {
+    pub summary: CheckpointSummary,
+    pub contents: CheckpointContents,
+    pub signatures: CheckpointSignatureMessage,
+}
+
+#[derive(Clone, Debug)]
+pub struct TransactionData {
+    pub transaction: Transaction,
+    pub effects: TransactionEffects,
+    pub events: Option<TransactionEvents>,
+    pub checkpoint_number: CheckpointSequenceNumber,
+    pub timestamp: u64,
```

Comment:
We found that a majority of the fetching of summaries was only to get a txns checkpoint number and timestamp so this could save a round trip at a small storage cost